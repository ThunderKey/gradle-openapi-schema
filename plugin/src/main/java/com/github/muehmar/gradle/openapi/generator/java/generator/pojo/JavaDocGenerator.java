package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;

public class JavaDocGenerator {
  private static final int MAX_LENGTH = 80;

  private JavaDocGenerator() {}

  public static <A, B> Generator<A, B> ofJavaDocString(String javaDoc) {
    return (a, b, writer) -> javaDoc().generate(javaDoc, (Void) null, writer);
  }

  public static Generator<String, Void> javaDoc() {
    return Generator.<String, Void>ofWriterFunction(w -> w.println("/**"))
        .append(content())
        .append(w -> w.println(" */"));
  }

  private static Generator<String, Void> content() {
    return (input, ign, writer) ->
        PList.fromArray(input.split("\n"))
            .flatMap(JavaDocGenerator::autoNewline)
            .map(line -> " * " + line)
            .foldLeft(writer, Writer::println);
  }

  private static PList<String> autoNewline(String line) {
    if (line.length() <= MAX_LENGTH) {
      return PList.single(line);
    }

    final int lastWhiteSpaceWithinMaxLength = line.substring(0, 80).lastIndexOf(" ");
    final int firstWhitespace = line.indexOf(" ");

    final int firstWhitespaceOrLineLength = firstWhitespace > 0 ? firstWhitespace : line.length();
    final int newLineIndex =
        lastWhiteSpaceWithinMaxLength > 0
            ? lastWhiteSpaceWithinMaxLength
            : firstWhitespaceOrLineLength;

    if (newLineIndex >= line.length() - 2) {
      return PList.single(line);
    }

    final String firstLine = line.substring(0, newLineIndex);
    final String remaining = line.substring(newLineIndex + 1);
    return autoNewline(remaining).cons(firstLine);
  }
}
