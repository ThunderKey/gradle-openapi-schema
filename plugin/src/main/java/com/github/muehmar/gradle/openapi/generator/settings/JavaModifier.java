package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import java.io.Serializable;
import java.util.Optional;

public enum JavaModifier implements Serializable {
  PRIVATE("private"),
  PROTECTED("protected"),
  PACKAGE_PRIVATE("package-private"),
  PUBLIC("public");

  private final String value;

  JavaModifier(String value) {
    this.value = value;
  }

  public String getValue() {
    return value;
  }

  public static Optional<JavaModifier> fromString(String value) {
    return PList.of(values()).find(modifier -> modifier.value.equalsIgnoreCase(value));
  }
}
