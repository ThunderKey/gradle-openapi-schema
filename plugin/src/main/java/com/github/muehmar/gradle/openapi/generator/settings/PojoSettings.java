package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojoextension.annotations.PojoExtension;
import java.io.Serializable;
import java.util.List;
import java.util.Objects;

@PojoExtension
public class PojoSettings extends PojoSettingsExtension implements Serializable {
  private final JsonSupport jsonSupport;
  private final String packageName;
  private final String suffix;
  private final boolean enableSafeBuilder;
  private final boolean enableConstraints;
  private final List<ClassTypeMapping> classTypeMappings;
  private final List<FormatTypeMapping> formatTypeMappings;
  private final EnumDescriptionSettings enumDescriptionSettings;

  PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      List<ClassTypeMapping> classTypeMappings,
      List<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings;
    this.formatTypeMappings = formatTypeMappings;
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      PList<ClassTypeMapping> classTypeMappings,
      PList<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings.toArrayList();
    this.formatTypeMappings = formatTypeMappings.toArrayList();
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public String getPackageName() {
    return packageName;
  }

  public String getSuffix() {
    return suffix;
  }

  public JsonSupport getJsonSupport() {
    return jsonSupport;
  }

  public boolean isJacksonJson() {
    return jsonSupport.equals(JsonSupport.JACKSON);
  }

  public boolean isEnableConstraints() {
    return enableConstraints;
  }

  public PList<ClassTypeMapping> getClassTypeMappings() {
    return PList.fromIter(classTypeMappings);
  }

  public PList<FormatTypeMapping> getFormatTypeMappings() {
    return PList.fromIter(formatTypeMappings);
  }

  public boolean isEnableSafeBuilder() {
    return enableSafeBuilder;
  }

  public boolean isDisableSafeBuilder() {
    return !isEnableSafeBuilder();
  }

  public EnumDescriptionSettings getEnumDescriptionSettings() {
    return enumDescriptionSettings;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    PojoSettings that = (PojoSettings) o;
    return enableSafeBuilder == that.enableSafeBuilder
        && enableConstraints == that.enableConstraints
        && jsonSupport == that.jsonSupport
        && Objects.equals(packageName, that.packageName)
        && Objects.equals(suffix, that.suffix)
        && Objects.equals(classTypeMappings, that.classTypeMappings)
        && Objects.equals(formatTypeMappings, that.formatTypeMappings)
        && Objects.equals(enumDescriptionSettings, that.enumDescriptionSettings);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
        jsonSupport,
        packageName,
        suffix,
        enableSafeBuilder,
        enableConstraints,
        classTypeMappings,
        formatTypeMappings,
        enumDescriptionSettings);
  }

  @Override
  public String toString() {
    return "PojoSettings{"
        + "jsonSupport="
        + jsonSupport
        + ", packageName='"
        + packageName
        + '\''
        + ", suffix='"
        + suffix
        + '\''
        + ", enableSafeBuilder="
        + enableSafeBuilder
        + ", enableConstraints="
        + enableConstraints
        + ", classTypeMappings="
        + classTypeMappings
        + ", formatTypeMappings="
        + formatTypeMappings
        + ", enumDescriptionExtraction="
        + enumDescriptionSettings
        + '}';
  }
}
