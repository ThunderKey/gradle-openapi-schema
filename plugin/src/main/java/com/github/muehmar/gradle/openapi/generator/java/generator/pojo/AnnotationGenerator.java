package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;

public class AnnotationGenerator {
  private AnnotationGenerator() {}

  public static <A, B> Generator<A, B> override() {
    return Generator.ofWriterFunction(w -> w.println("@Override"));
  }

  public static <A> Generator<A, PojoSettings> deprecatedRawGetter() {
    return Generator.<A, PojoSettings>ofWriterFunction(w -> w.println("@Deprecated"))
        .filter((ignore, settings) -> settings.getRawGetter().isDeprecatedAnnotation());
  }
}
