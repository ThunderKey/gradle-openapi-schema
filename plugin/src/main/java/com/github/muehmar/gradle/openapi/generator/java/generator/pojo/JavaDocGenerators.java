package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import com.github.muehmar.gradle.openapi.generator.java.generator.shared.JavaDocGenerator;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;

public class JavaDocGenerators {
  private JavaDocGenerators() {}

  public static <A> Generator<A, PojoSettings> deprecatedValidationMethodJavaDoc() {
    return JavaDocGenerator.<A, PojoSettings>ofJavaDocString(
            "@deprecated This method is intended to be used only by frameworks for validation or serialisation.")
        .filter((a, settings) -> settings.getValidationMethods().isDeprecatedAnnotation());
  }
}
