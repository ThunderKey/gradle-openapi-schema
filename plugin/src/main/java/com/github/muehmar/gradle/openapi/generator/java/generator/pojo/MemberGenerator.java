package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties.additionalPropertiesName;
import static io.github.muehmar.codegenerator.Generator.constant;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.JavaRefs;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.jackson.JacksonAnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Optional;
import lombok.Value;

public class MemberGenerator {
  private MemberGenerator() {}

  public static Generator<MemberContent, PojoSettings> memberGenerator() {
    return Generator.<MemberContent, PojoSettings>emptyGen()
        .append(jsonValueAnnotation())
        .appendList(singleMember(), MemberContent::getMembers)
        .appendOptional(additionalPropertiesMember(), MemberContent::getAdditionalProperties);
  }

  private static Generator<JavaPojoMember, PojoSettings> singleMember() {
    final Generator<JavaPojoMember, PojoSettings> fieldDeclaration =
        (field, settings, writer) ->
            writer.println(
                "private final %s %s;",
                field.getJavaType().getFullClassName(), field.getNameAsIdentifier());
    final Generator<JavaPojoMember, PojoSettings> requiredNullableFlag =
        (field, settings, writer) ->
            writer.println("private final boolean %s;", field.getIsPresentFlagName());
    final Generator<JavaPojoMember, PojoSettings> optionalNullableFlag =
        (field, settings, writer) ->
            writer.println("private final boolean %s;", field.getIsNullFlagName());
    return fieldDeclaration
        .appendConditionally(JavaPojoMember::isRequiredAndNullable, requiredNullableFlag)
        .appendConditionally(JavaPojoMember::isOptionalAndNullable, optionalNullableFlag)
        .append(
            (field, settings, writer) ->
                field.getJavaType().getImportsAsString().foldLeft(writer, Writer::ref));
  }

  private static Generator<JavaAdditionalProperties, PojoSettings> additionalPropertiesMember() {
    return Generator.<JavaAdditionalProperties, PojoSettings>emptyGen()
        .append(constant("private final Map<String, Object> %s;", additionalPropertiesName()))
        .append(RefsGenerator.ref(JavaRefs.JAVA_UTIL_MAP));
  }

  private static Generator<MemberContent, PojoSettings> jsonValueAnnotation() {
    return JacksonAnnotationGenerator.<MemberContent>jsonValue().filter(MemberContent::isArrayPojo);
  }

  @Value
  @PojoBuilder(builderName = "MemberContentBuilder")
  public static class MemberContent {
    boolean isArrayPojo;
    PList<JavaPojoMember> members;
    Optional<JavaAdditionalProperties> additionalProperties;
  }
}
