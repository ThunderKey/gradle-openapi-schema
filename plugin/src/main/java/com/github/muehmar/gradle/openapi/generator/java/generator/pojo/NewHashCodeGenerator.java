package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import static com.github.muehmar.gradle.openapi.util.Functions.allExceptFirst;
import static io.github.muehmar.codegenerator.java.JavaModifier.PUBLIC;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.JavaRefs;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojo;
import com.github.muehmar.gradle.openapi.generator.model.Name;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.JavaGenerators;
import io.github.muehmar.codegenerator.writer.Writer;

public class NewHashCodeGenerator {
  private NewHashCodeGenerator() {}

  public static Generator<JavaPojo, PojoSettings> hashCodeMethod() {
    final Generator<JavaPojo, PojoSettings> method =
        JavaGenerators.<JavaPojo, PojoSettings>methodGen()
            .modifiers(PUBLIC)
            .noGenericTypes()
            .returnType("int")
            .methodName("hashCode")
            .noArguments()
            .content(hashCodeMethodContent())
            .build();
    return AnnotationGenerator.<JavaPojo, PojoSettings>override()
        .append(method)
        .filter(pojo -> pojo.fold(arrayPojo -> true, enumPojo -> false, objectPojo -> true));
  }

  private static Generator<JavaPojo, PojoSettings> hashCodeMethodContent() {
    return (pojo, s, w) -> {
      final PList<String> fieldNames =
          pojo.getMembersOrEmpty()
              .flatMap(
                  member -> {
                    final Name memberName = member.getName();
                    if (member.isRequiredAndNullable()) {
                      final String requiredNullableFlagName =
                          String.format("is%sPresent", memberName.startUpperCase());
                      return PList.of(memberName.asString(), requiredNullableFlagName);
                    } else if (member.isOptionalAndNullable()) {
                      final String optionalNullableFlagName =
                          String.format("is%sNull", memberName.startUpperCase());
                      return PList.of(memberName.asString(), optionalNullableFlagName);
                    } else {
                      return PList.single(memberName.asString());
                    }
                  });

      final Writer writerStartPrinted = w.println("return Objects.hash(");

      final PList<String> mappedFieldNames =
          fieldNames.reverse().zipWithIndex().map(allExceptFirst(name -> name + ",")).reverse();

      return mappedFieldNames
          .foldLeft(writerStartPrinted, (writer, name) -> writer.tab(1).println(name))
          .println(");")
          .ref(JavaRefs.JAVA_UTIL_OBJECTS);
    };
  }
}
