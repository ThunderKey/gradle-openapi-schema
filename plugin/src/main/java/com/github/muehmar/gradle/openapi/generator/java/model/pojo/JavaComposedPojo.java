package com.github.muehmar.gradle.openapi.generator.java.model.pojo;

import static com.github.muehmar.gradle.openapi.generator.model.pojo.ComposedPojo.CompositionType.ANY_OF;
import static com.github.muehmar.gradle.openapi.generator.model.pojo.ComposedPojo.CompositionType.ONE_OF;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.model.Discriminator;
import com.github.muehmar.gradle.openapi.generator.model.PojoName;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.pojo.ComposedPojo;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class JavaComposedPojo implements JavaPojo {
  private final PojoName name;
  private final String description;
  private final PList<JavaPojo> javaPojos;
  private final ComposedPojo.CompositionType compositionType;
  private final Constraints constraints;
  private final Optional<Discriminator> discriminator;

  JavaComposedPojo(
      PojoName name,
      String description,
      PList<JavaPojo> javaPojos,
      ComposedPojo.CompositionType compositionType,
      Constraints constraints,
      Optional<Discriminator> discriminator) {
    this.name = name;
    this.description = Optional.ofNullable(description).orElse("");
    this.javaPojos = javaPojos;
    this.compositionType = compositionType;
    this.constraints = constraints;
    this.discriminator = discriminator;
  }

  public static JavaComposedPojo wrap(ComposedPojo composedPojo, TypeMappings typeMappings) {
    return new JavaComposedPojo(
        composedPojo.getName(),
        composedPojo.getDescription(),
        composedPojo.getPojos().map(pojo -> JavaPojo.wrap(pojo, typeMappings)),
        composedPojo.getCompositionType(),
        composedPojo.getConstraints(),
        composedPojo.getDiscriminator());
  }

  @Override
  public PojoName getName() {
    return name;
  }

  @Override
  public String getDescription() {
    return description;
  }

  public PList<JavaPojo> getJavaPojos() {
    return javaPojos;
  }

  public ComposedPojo.CompositionType getCompositionType() {
    return compositionType;
  }

  public boolean isAnyOf() {
    return compositionType.equals(ANY_OF);
  }

  public boolean isOneOf() {
    return compositionType.equals(ONE_OF);
  }

  public Optional<Discriminator> getDiscriminator() {
    return discriminator;
  }

  public JavaObjectPojo wrapIntoJavaObjectPojo() {
    return JavaObjectPojo.from(name, description, getMembers(), constraints);
  }

  public PList<JavaPojoMember> getMembers() {
    return javaPojos.flatMap(JavaPojo::getMembersOrEmpty).distinct(Function.identity());
  }

  @Override
  public <T> T fold(
      Function<JavaArrayPojo, T> onArrayPojo,
      Function<JavaEnumPojo, T> onEnumPojo,
      Function<JavaObjectPojo, T> onObjectPojo,
      Function<JavaComposedPojo, T> onComposedPojo,
      Function<JavaFreeFormPojo, T> onFreeFormPojo) {
    return onComposedPojo.apply(this);
  }
}
