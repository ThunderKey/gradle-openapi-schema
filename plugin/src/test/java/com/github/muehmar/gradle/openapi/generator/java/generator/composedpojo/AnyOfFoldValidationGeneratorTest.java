package com.github.muehmar.gradle.openapi.generator.java.generator.composedpojo;

import static com.github.muehmar.gradle.openapi.generator.model.pojo.ComposedPojo.CompositionType.ANY_OF;
import static com.github.muehmar.gradle.openapi.generator.model.pojo.ComposedPojo.CompositionType.ONE_OF;
import static org.junit.jupiter.api.Assertions.*;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaComposedPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.settings.JavaModifier;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith(SnapshotExtension.class)
class AnyOfFoldValidationGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("AnyOf")
  void generate_when_anyOf_then_correctOutput() {
    final Generator<JavaComposedPojo, PojoSettings> generator =
        AnyOfFoldValidationGenerator.generator();
    final Writer writer =
        generator.generate(
            JavaPojos.composedPojo(ANY_OF),
            TestPojoSettings.defaultSettings(),
            Writer.createDefault());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("AnyOfProtectedAndDeprecatedSettings")
  void generate_when_anyOfProtectedAndDeprecatedSettings_then_correctOutput() {
    final Generator<JavaComposedPojo, PojoSettings> generator =
        AnyOfFoldValidationGenerator.generator();
    final Writer writer =
        generator.generate(
            JavaPojos.composedPojo(ANY_OF),
            TestPojoSettings.defaultSettings()
                .withValidationMethods(
                    TestPojoSettings.defaultValidationMethods()
                        .withModifier(JavaModifier.PROTECTED)
                        .withDeprecatedAnnotation(true)),
            Writer.createDefault());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  void generate_when_oneOf_then_noOutput() {
    final Generator<JavaComposedPojo, PojoSettings> generator =
        AnyOfFoldValidationGenerator.generator();
    final Writer writer =
        generator.generate(
            JavaPojos.composedPojo(ONE_OF),
            TestPojoSettings.defaultSettings(),
            Writer.createDefault());

    assertEquals("", writer.asString());
  }
}
