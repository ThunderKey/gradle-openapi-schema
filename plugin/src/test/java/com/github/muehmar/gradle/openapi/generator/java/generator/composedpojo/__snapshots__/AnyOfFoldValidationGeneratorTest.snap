AnyOf=[
@Valid
private List<Object> getAnyOf() {
  if (getValidCount() == 0) {
    return null;
  }
  return fold(userdto -> userdto, necessityandnullabilitydto -> necessityandnullabilitydto);
}
]